xmodule/services.py

```diff
@@ -308,3 +312,112 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ResultService(Service):
+    """
+    An XBlock Service that allows XModules to define:
+
+    - if the answers are visible
+    - whether correctness is currently hidden. When correctness is hidden, this
+    limits the user's access to the correct/incorrect flags, messages, problem
+    scores, and aggregate subsection and course grades.
+    """
+    def __init__(self, block=None, user_is_staff=False, **kwargs):
+        super().__init__(**kwargs)
+        self._user_is_staff = user_is_staff
+        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
+        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
+        self._xblock = self.xblock() or block
+
+    def answer_available(self):
+        """
+        Returns whether correctness is available now, for the given attributes.
+        """
+        if self._xblock:
+            show_correctness = self._xblock.show_correctness
+            show_answer = self._xblock.showanswer
+            past_due = self._xblock.is_past_due()
```

Comment:
Can you move the logic for these helpers into the service as well?

The goal is to separate out the things that could be common across many XBlocks, that are based on a simple subset of fields like calculations of attempts, due dates, etc.

The things that only an XBlock could determine for itself (like `is_correct`) would be passed in as parameters.

So the XBlock snippet calling the service would look something like:

```python
  show_answer = result_service.answer_available(self, self.is_correct())
```

With the idea being that we would eventually have a mixin that defines all the fields that this service needs in order to make determinations like whether something is past due. So the service does most of its logic based on those values (comparing dates and such). If something looks difficult to extract from `ProblemBlock` because it makes a call to `self.lcp`, pass it in as an explicit parameter to the service methods. The service code can assume certain XBlock fields exist (because they'll eventually be part of an agreed-upon mixin class of fields), but it should not assume that the XBlock implements any particular methods.

Reply:
I'm a bit confused on how to address this comment while preserving the changes asked in a previous review by @Agrendalath (see https://github.com/openedx/edx-platform/pull/33424#discussion_r1369006678). 

Do we want to pass parameters to the service about the XBlock? I think it is just a syntactic change because right now we are creating a new instance of the service for a given XBlock, which means it has access to all the information about itself without the need of adding additional parameters to the service methods.

Reply:
@Agrendalath's convinced me that we can have the service can read fields off of the XBlock for those fields that we know will be there from the InheritanceMixin. But I would still like to extract method logic as much as possible. For instance, `is_past_due` is a method on the ProblemBlock, but that's part of the logic that I would like to have extracted into this common runtime service.